util/coverage/common.py

```diff
@@ -0,0 +1,279 @@
+    run(BAZEL, "build", f"--config={config}", *device_libs)
+    obj_files = run(
+        BAZEL, "cquery", f"--config={config}", f"({' + '.join(device_libs)})",
+        "--output=starlark",
+        ("--starlark:expr='\\n'.join("
+         "[f.path for f in target.output_groups.compilation_outputs.to_list()]"
+         ")"))
+    log.info(f"object files with coverage data: {pformat(obj_files)}")
+    return obj_files
+
+
+def get_test_log_dirs(test_targets: List[str]) -> List[Path]:
+    """Get log direactories for the given test targets.
+
+    Args:
+        test_targets: Test targets.
+
+    Returns:
+        Log directories of the the given test targets.
+    """
+    [test_log_dir] = run(
+        BAZEL,
+        "info",
+        "bazel-testlogs",
+    )
+    return [Path(test_log_dir) / t[2:].replace(':', '/') for t in test_targets]
+
+
+def generate_report(out_dir: Path, merged_profile: Path, merged_library: Path,
+                    report_title: str, print_text_report: bool) -> None:
+    """Generate a coverage report.
+
+    This function generates a coverage report from the given merged profile and library.
+
+    Args:
+        out_dir: Path where to save the html report.
+        merged_profile: Path of merged profile data.
+        merged_library: Path of the merged library with coverage data.
+        print_text_report: Whether to print the text report.
+    """
+    # Generate html coverage report.
+    run(LLVM_COV, "show", "--show-line-counts", "--show-regions",
+        f"--project-title={report_title}", "--format=html",
+        f"--output-dir=./{out_dir}", "--instr-profile", str(merged_profile),
+        str(merged_library))
+    # Generate text coverage report.
+    with (out_dir / "report.txt").open("w") as f:
+        f.write("\n".join(
+            run(LLVM_COV, "report", "--instr-profile", str(merged_profile),
+                str(merged_library))))
+    if print_text_report:
+        print("\n".join(
+            run(LLVM_COV, "report", "--use-color", "--instr-profile",
+                str(merged_profile), str(merged_library))))
+    print(f"Saved coverage artifacts in {out_dir}")
+
+
+# Query for unit test targets
+TEST_TARGETS_INC = [
+    "//sw/device/silicon_creator/...",
+]
+TEST_TARGETS_EXC = [
+    "//sw/device/silicon_creator/rom_ext/...",
+    "attr(tags, 'dv', //sw/device/silicon_creator/...)",
+    "attr(tags, 'verilator', //sw/device/silicon_creator/...)",
+    "attr(tags, 'broken', //sw/device/silicon_creator/...)",
+]
+
+
+def test_targets_query(test_type: BazelTestType) -> str:
+    return (f"kind({test_type}, ({' + '.join(TEST_TARGETS_INC)})"
+            f" - ({' + '.join(TEST_TARGETS_EXC)}))")
+
+
+CoverageParams = namedtuple("CoverageParams", [
+    "config",
+    "libs_fn",
+    "objs_fn",
+    "test_targets_fn",
+    "test_log_dirs_fn",
+    "bazel_test_type",
+    "report_title",
+])
+
+
+def measure_coverage(*, log_level: LogLevel, out_root_dir: Path,
```

Comment:
The bare `*` parameter was new to me. This stackoverflow post indicates that forces the caller to used named parameters: <https://stackoverflow.com/a/14302007>.

I don't often resort to OOP, but I think it might make the most sense here. You could define `class AbstractCoverageReport` with abstract `get_libs()`, `get_objs()`, etc. and define subclasses as needed. That would reduce the number of params and (subjectively) make control flow easier to follow.

Reply:
I thought about defining a class but I thought it would be an overkill for a few callbacks. I'm not opposed to it though, except it can probably happen after (1) finishing the remaining parts (func and e2e test support) and (2) seeing how much of this can be moved to bazel.